gossip/minimal/goshawk.go

```diff
@@ -56,6 +61,132 @@ type FetchOptions struct {
 	BatchSize int
 	// Number of concurrent fetcher workers to run.
 	ParallelFetch int
+	// Manage hub retrieval state persistence.
+	State         ScanStateManager
+	FlushInterval time.Duration
+}
+
+// ScanStateManager controls hub scanning state, with the intention of allowing
+// scanning to resume where it was left off across restarts etc.
+type ScanStateManager interface {
+	// GetHubIndex returns the index at which scanning should commence for the given hub URL.
+	GetHubIndex(hubURL string) int64
+	// UpdateHubIndex indicates that future scanning of the given hub should start from at least nextIndex.
+	UpdateHubIndex(hubURL string, nextIndex int64)
+	// Flush ensures state is stored.
+	Flush(ctx context.Context) error
+}
+
+// ScanState holds a (mutex-protected) map of hub URL to next index to scan from.
+type ScanState struct {
+	Mu   sync.Mutex
+	Next map[string]int64
```

Comment:
Possible use case for sync.Map? Their doc always puts me off whenever I think about using it though.